### PR TITLE
Exclude

### PR DESCRIPTION
**Pull request by raymondhardynike**
This allows a user to exclude files when using dir archive.

### CODE REVIEW

**Comment by Constantin07** [comment]
I also need to exclude some files from directory before zipping it.
Any idea when this PR will be approved ?

File: archive/data_source_archive_file_test.go [hunk 1]
```diff
@@ -115,6 +122,15 @@ data "archive_file" "foo" {
 }
 `
 
+var testAccArchiveFileDirExcludesConfig = `
+data "archive_file" "foo" {
+  type        = "zip"
+  source_dir  = "../archive/test-fixtures/../test-fixtures/test-dir"
+	excludes    = ["test-fixtures/test-dir/file2.txt"]
```
Comment (paultyng) [inline archive/data_source_archive_file_test.go]: some whitespace weirdness in here

File: archive/zip_archiver.go [hunk 2]
```diff
@@ -83,12 +98,25 @@ func (a *ZipArchiver) ArchiveDir(indirname string) error {
 	defer a.close()
 
 	return filepath.Walk(indirname, func(path string, info os.FileInfo, err error) error {
+		tmpFilePath, _ := filepath.Abs(path)
+		tmpLongFileName, _ := filepath.Abs(indirname)
+		shortFileName := strings.TrimPrefix(tmpFilePath, tmpLongFileName)
```
Comment (paultyng) [inline archive/zip_archiver.go]: Please use `filepath.Rel` (it's already used below, you could just move that line up)

File: archive/zip_archiver.go [hunk 3]
```diff
@@ -83,12 +98,25 @@ func (a *ZipArchiver) ArchiveDir(indirname string) error {
...
+		isMatch := checkMatch(shortFileName, excludes)
 		if info.IsDir() {
+			if isMatch {
```
Comment (paultyng) [inline archive/zip_archiver.go]: I think you probably still want to walk all the dirs and match against the full relative paths? You may want to add a test case that covers this.

Comment (raymondhardynike) [reply archive/zip_archiver.go]: If you match on a directory you are excluding the directory why match after that?

i.e.
/hi/
/hello/world
/hello/i/am/here
if you exclude hello/ you should exclude everything in the hello path why still walk?

File: archive/zip_archiver.go (hunk 2 above)
Comment (raymondhardynike) [reply archive/zip_archiver.go]: I did try to use filepath.Rel at first this causes some problems when matching. The reason is that Rel path could look like this
/somedir/moredir/evenmoredir/../hi/
When trying to remove the prefix from that the prefix would just be
/somedir/moredir/hi/

File: archive/data_source_archive_file_test.go (hunk 1 above)
Comment (raymondhardynike) [reply archive/data_source_archive_file_test.go]: I can fix this

**Comment by raymondhardynike** [comment]
I have updated this to use relative path for exact match removing globing on anything except dirs.

File: archive/data_source_archive_file.go [hunk 4]
```diff
@@ -77,6 +77,15 @@ func dataSourceFile() *schema.Resource {
 				ForceNew:      true,
 				ConflictsWith: []string{"source_content", "source_content_filename", "source_file"},
 			},
+			"excludes": &schema.Schema{
+				Type:          schema.TypeList,
```
[the remark this answers is not in the document]
Comment (raymondhardynike) [reply archive/data_source_archive_file.go]: Yeah typeSet works

**Pull request merged**